### ISSUE TITLE
`restic backup` hangs indefinitely 

### ISSUE THREAD

**Issue by mifix** [restic/restic#3020]
Output of `restic version`
--------------------------
`restic 0.10.0 compiled with go1.15.2 on linux/amd64`

How did you run restic exactly?
-------------------------------
```
restic backup --verbose=5 --one-file-system --option b2.connections=10 --exclude-file /etc/restic/backup_exclude / /srv /home /srv/dev-disk-by-label-data01
```

What backend/server/service did you use to store the repository?
----------------------------------------------------------------
Backblaze B2

Expected behavior
-----------------
I'd expect `restic backup` to finish and exit.

Actual behavior
---------------
```bash
$ restic backup ....
unchanged ....
[....]
unchanged /srv/
[3:12] 482117 files 658.586 GiB, total 840720 files 914.373 GiB, 0 errors   # <---- HANGS HERE
```

Steps to reproduce the behavior
-------------------------------

Happens every time I run restic backup, except the inital backup worked.

Do you have any idea what may have caused this?
-----------------------------------------------
No.

Do you have an idea how to solve the issue?
-------------------------------------------
No. Maybe a similar issue: #1526

Output of `strace -f -p $(pidof restic)`:
```
[pid 23718] <... epoll_pwait resumed> [{EPOLLIN, {u32=22792272, u64=22792272}}], 128, 8190, NULL, 0) = 1
[pid 23721] futex(0xc000051148, FUTEX_WAIT_PRIVATE, 0, NULL <unfinished ...>
[pid 23718] read(5,  <unfinished ...>
[pid 23713] <... nanosleep resumed> NULL) = 0
[pid 23718] <... read resumed> "\0", 16) = 1
[pid 23713] futex(0x158bad8, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=50201594} <unfinished ...>
[pid 23718] epoll_pwait(4,  <unfinished ...>
[pid 23712] futex(0x158c588, FUTEX_WAIT_PRIVATE, 0, NULL <unfinished ...>
[pid 23718] <... epoll_pwait resumed> [], 128, 0, NULL, 6967186) = 0
[pid 23718] epoll_pwait(4,  <unfinished ...>
[pid 23713] <... futex resumed> )       = -1 ETIMEDOUT (Connection timed out)
[pid 23718] <... epoll_pwait resumed> [], 128, 50, NULL, 0) = 0
[pid 23713] futex(0xc000051148, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid 23718] epoll_pwait(4,  <unfinished ...>
[pid 23721] <... futex resumed> )       = 0
[pid 23718] <... epoll_pwait resumed> [], 128, 0, NULL, 6967200) = 0
[pid 23721] futex(0x158c588, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid 23718] futex(0xc000050548, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid 23721] <... futex resumed> )       = 1
[pid 23718] <... futex resumed> )       = 1
[pid 23721] epoll_pwait(4,  <unfinished ...>
[pid 23718] madvise(0xc006400000, 2097152, MADV_NOHUGEPAGE <unfinished ...>
[pid 23721] <... epoll_pwait resumed> [], 128, 0, NULL, 0) = 0
[pid 23718] <... madvise resumed> )     = 0
[pid 23721] epoll_pwait(4,  <unfinished ...>
[pid 23718] madvise(0xc006458000, 8192, MADV_FREE <unfinished ...>
[pid 23714] <... futex resumed> )       = 0
[pid 23718] <... madvise resumed> )     = 0
[pid 23714] futex(0xc000050548, FUTEX_WAIT_PRIVATE, 0, NULL <unfinished ...>
[pid 23718] write(6, "\0", 1 <unfinished ...>
[pid 23713] <... futex resumed> )       = 1
[pid 23721] <... epoll_pwait resumed> [{EPOLLIN, {u32=22792272, u64=22792272}}], 128, 8138, NULL, 0) = 1
[pid 23718] <... write resumed> )       = 1
[pid 23721] read(5,  <unfinished ...>
[pid 23718] futex(0xc000108148, FUTEX_WAIT_PRIVATE, 0, NULL <unfinished ...>
[pid 23721] <... read resumed> "\0", 16) = 1
[pid 23713] nanosleep({tv_sec=0, tv_nsec=20000},  <unfinished ...>
[pid 23721] epoll_pwait(4,  <unfinished ...>
[pid 23712] <... futex resumed> )       = 0
[pid 23721] <... epoll_pwait resumed> [], 128, 0, NULL, 6967200) = 0
[pid 23713] <... nanosleep resumed> NULL) = 0
[pid 23721] epoll_pwait(4,  <unfinished ...>
[pid 23713] nanosleep({tv_sec=0, tv_nsec=20000},  <unfinished ...>
[pid 23712] epoll_pwait(4, [], 128, 0, NULL, 6967200) = 0
[pid 23713] <... nanosleep resumed> NULL) = 0
[pid 23712] futex(0x158c588, FUTEX_WAIT_PRIVATE, 0, NULL <unfinished ...>
[pid 23713] futex(0x158bad8, FUTEX_WAIT_PRIVATE, 0, {tv_sec=0, tv_nsec=42678707}
```

Output of restic after SIGQUIT (^+\):
```
^\SIGQUIT: quit
PC=0x46ce81 m=0 sigcode=128

goroutine 0 [idle]:
runtime.futex(0x158c588, 0x80, 0x0, 0x0, 0x7ffc00000000, 0x46ca5c, 0xb18c, 0x12a9071, 0x7ffcc1049268, 0x40beff, ...)
        /usr/local/go/src/runtime/sys_linux_amd64.s:587 +0x21
runtime.futexsleep(0x158c588, 0x0, 0xffffffffffffffff)
        /usr/local/go/src/runtime/os_linux.go:45 +0x46
runtime.notesleep(0x158c588)
        /usr/local/go/src/runtime/lock_futex.go:159 +0x9f
runtime.stopm()
        /usr/local/go/src/runtime/proc.go:1910 +0xc5
runtime.findrunnable(0xc000043000, 0x0)
        /usr/local/go/src/runtime/proc.go:2471 +0xa7f
runtime.schedule()
        /usr/local/go/src/runtime/proc.go:2669 +0x2d7
runtime.park_m(0xc000000c00)
        /usr/local/go/src/runtime/proc.go:2837 +0x9d
runtime.mcall(0x0)
        /usr/local/go/src/runtime/asm_amd64.s:318 +0x5b

goroutine 1 [chan receive, 5 minutes]:
gopkg.in/tomb%2ev2.(*Tomb).Wait(0xc0000cab40, 0x0, 0x0)
        /home/build/go/pkg/mod/gopkg.in/tomb.v2@v2.0.0-20161208151619-d5d1b5820637/tomb.go:126 +0x4a
main.runBackup(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc0001b9700, ...)
        /restic/cmd/restic/cmd_backup.go:616 +0x1156
main.glob..func2(0x1579c60, 0xc0002b2300, 0x4, 0xc, 0x0, 0x0)
        /restic/cmd/restic/cmd_backup.go:70 +0x1ba
github.com/spf13/cobra.(*Command).execute(0x1579c60, 0xc0002b2240, 0xc, 0xc, 0x1579c60, 0xc0002b2240)
        /home/build/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:842 +0x47c
github.com/spf13/cobra.(*Command).ExecuteC(0x157ce40, 0xc0001b9600, 0x4, 0x10)
        /home/build/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
        /home/build/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887
main.main()
        /restic/cmd/restic/main.go:86 +0x3f

goroutine 18 [chan receive, 8 minutes]:
github.com/restic/restic/internal/restic.init.0.func1.1()
        /restic/internal/restic/lock.go:258 +0xb3
created by github.com/restic/restic/internal/restic.init.0.func1
        /restic/internal/restic/lock.go:255 +0x35

goroutine 20 [syscall, 8 minutes]:
os/signal.signal_recv(0x0)
        /usr/local/go/src/runtime/sigqueue.go:147 +0x9d
os/signal.loop()
        /usr/local/go/src/os/signal/signal_unix.go:23 +0x25
created by os/signal.Notify.func1.1
        /usr/local/go/src/os/signal/signal.go:150 +0x45

goroutine 21 [chan receive, 8 minutes]:
github.com/restic/restic/internal/restic.init.2.func1(0xc0000b02a0)
        /restic/internal/restic/progress_unix.go:17 +0x65
created by github.com/restic/restic/internal/restic.init.2
        /restic/internal/restic/progress_unix.go:16 +0xab

goroutine 33 [select]:
go.opencensus.io/stats/view.(*worker).start(0xc000210380)
        /home/build/go/pkg/mod/go.opencensus.io@v0.22.4/stats/view/worker.go:276 +0x105
created by go.opencensus.io/stats/view.init.0
        /home/build/go/pkg/mod/go.opencensus.io@v0.22.4/stats/view/worker.go:34 +0x68

goroutine 34 [chan receive, 8 minutes]:
main.CleanupHandler(0xc00020ccc0)
        /restic/cmd/restic/cleanup.go:59 +0x4c
created by main.init.0
        /restic/cmd/restic/cleanup.go:21 +0x67

goroutine 23 [select, 5 minutes]:
github.com/restic/restic/internal/ui/termstatus.(*Terminal).run(0xc0000b17a0, 0x103ef00, 0xc000300140)
        /restic/internal/ui/termstatus/status.go:96 +0x165
github.com/restic/restic/internal/ui/termstatus.(*Terminal).Run(0xc0000b17a0, 0x103ef00, 0xc000300140)
        /restic/internal/ui/termstatus/status.go:85 +0x70
main.glob..func2.1(0x0, 0x0)
        /restic/cmd/restic/cmd_backup.go:68 +0x6d
gopkg.in/tomb%2ev2.(*Tomb).run(0xc0000caaf0, 0xc0000c1f20)
        /home/build/go/pkg/mod/gopkg.in/tomb.v2@v2.0.0-20161208151619-d5d1b5820637/tomb.go:163 +0x38
created by gopkg.in/tomb%2ev2.(*Tomb).Go
        /home/build/go/pkg/mod/gopkg.in/tomb.v2@v2.0.0-20161208151619-d5d1b5820637/tomb.go:159 +0xbc

goroutine 97 [select, 5 minutes]:
main.refreshLocks(0x158a8a0, 0xc00012e2a0)
        /restic/cmd/restic/lock.go:70 +0x1c5
created by main.lockRepository
        /restic/cmd/restic/lock.go:47 +0x26c

goroutine 105 [select, 5 minutes]:
github.com/restic/restic/internal/ui.(*Backup).ReportTotal(0xc0001b65b0, 0xc00533ba20, 0xa5, 0xcd45a, 0x33c99, 0x2604, 0xe497d7da48)
        /restic/internal/ui/backup.go:343 +0xf7
github.com/restic/restic/internal/archiver.(*Scanner).scan(0xc0005db320, 0x103ef00, 0xc0000b2200, 0xcd45a, 0x33c99, 0x2604, 0xe497d7da48, 0xc00533ba20, 0xa5, 0xcd45a, ..
.)
        /restic/internal/archiver/scanner.go:107 +0x1f0
github.com/restic/restic/internal/archiver.(*Scanner).scan(0xc0005db320, 0x103ef00, 0xc0000b2200, 0xcd45a, 0x33c98, 0x2604, 0xe497d7da48, 0xc00500b540, 0xa0, 0xcd452, ..
.)
        /restic/internal/archiver/scanner.go:97 +0x4a5
github.com/restic/restic/internal/archiver.(*Scanner).scan(0xc0005db320, 0x103ef00, 0xc0000b2200, 0xcd452, 0x33c90, 0x2604, 0xe497d7cd64, 0xc00500afa0, 0x92, 0xcd409, ..
.)
        /restic/internal/archiver/scanner.go:97 +0x4a5
github.com/restic/restic/internal/archiver.(*Scanner).scan(0xc0005db320, 0x103ef00, 0xc0000b2200, 0xcd409, 0x33c8d, 0x2604, 0xe497d743d2, 0xc00a788fc0, 0x8e, 0x0, ...)
        /restic/internal/archiver/scanner.go:97 +0x4a5
github.com/restic/restic/internal/archiver.(*Scanner).scan(0xc0005db320, 0x103ef00, 0xc0000b2200, 0xcc455, 0x33af3, 0x2604, 0xe497b182fc, 0xc00a788e10, 0x87, 0x0, ...)
        /restic/internal/archiver/scanner.go:97 +0x4a5
github.com/restic/restic/internal/archiver.(*Scanner).scan(0xc0005db320, 0x103ef00, 0xc0000b2200, 0xcc455, 0x33af3, 0x2604, 0xe497b182fc, 0xc00a788c60, 0x81, 0xcc455, ..
.)
        /restic/internal/archiver/scanner.go:97 +0x4a5
github.com/restic/restic/internal/archiver.(*Scanner).scan(0xc0005db320, 0x103ef00, 0xc0000b2200, 0xcc455, 0x33af3, 0x2604, 0xe497b182fc, 0xc006c7ca00, 0x7e, 0xcc450, ..
.)
        /restic/internal/archiver/scanner.go:97 +0x4a5
github.com/restic/restic/internal/archiver.(*Scanner).scan(0xc0005db320, 0x103ef00, 0xc0000b2200, 0xcc450, 0x33aef, 0x2604, 0xe497b10f5b, 0xc005cd8000, 0x78, 0xcb10e, ..
.)
        /restic/internal/archiver/scanner.go:97 +0x4a5
github.com/restic/restic/internal/archiver.(*Scanner).scan(0xc0005db320, 0x103ef00, 0xc0000b2200, 0xcb10e, 0x337f6, 0x2289, 0xe493549047, 0xc0039cf880, 0x72, 0xb6d5e, ..
.)
        /restic/internal/archiver/scanner.go:97 +0x4a5
github.com/restic/restic/internal/archiver.(*Scanner).scan(0xc0005db320, 0x103ef00, 0xc0000b2200, 0xb6d5e, 0x302b0, 0x1b92, 0xe479bc8c7d, 0xc006be6620, 0x69, 0xb6b93, ..
.)
        /restic/internal/archiver/scanner.go:97 +0x4a5
github.com/restic/restic/internal/archiver.(*Scanner).scan(0xc0005db320, 0x103ef00, 0xc0000b2200, 0xb6b93, 0x30259, 0x1b91, 0xe479a628e2, 0xc001d723f0, 0x62, 0x0, ...)
        /restic/internal/archiver/scanner.go:97 +0x4a5
github.com/restic/restic/internal/archiver.(*Scanner).scan(0xc0005db320, 0x103ef00, 0xc0000b2200, 0xb41d9, 0x2fb51, 0x1b8b, 0xe47141bc43, 0xc0061fa840, 0x5c, 0x0, ...)
        /restic/internal/archiver/scanner.go:97 +0x4a5
github.com/restic/restic/internal/archiver.(*Scanner).scan(0xc0005db320, 0x103ef00, 0xc0000b2200, 0xb41d9, 0x2fb51, 0x1b8b, 0xe47141bc43, 0xc0061fa720, 0x56, 0xb41d9, ..
.)
        /restic/internal/archiver/scanner.go:97 +0x4a5
github.com/restic/restic/internal/archiver.(*Scanner).scan(0xc0005db320, 0x103ef00, 0xc0000b2200, 0xb41d9, 0x2fb51, 0x1b8b, 0xe47141bc43, 0xc00663f2c0, 0x51, 0xb35d2, ..
.)
        /restic/internal/archiver/scanner.go:97 +0x4a5
github.com/restic/restic/internal/archiver.(*Scanner).scan(0xc0005db320, 0x103ef00, 0xc0000b2200, 0xb35d2, 0x2f94d, 0x1b8b, 0xe4115cb2db, 0xc003d2c500, 0x46, 0xb353b, ..
.)
        /restic/internal/archiver/scanner.go:97 +0x4a5
github.com/restic/restic/internal/archiver.(*Scanner).scan(0xc0005db320, 0x103ef00, 0xc0000b2200, 0xb353b, 0x2f945, 0x1b8b, 0xe410fe9056, 0xc0048880c0, 0x36, 0x8fe91, ..
.)
        /restic/internal/archiver/scanner.go:97 +0x4a5
github.com/restic/restic/internal/archiver.(*Scanner).scan(0xc0005db320, 0x103ef00, 0xc0000b2200, 0x8fe91, 0x239d9, 0x1b7f, 0xb12bd203fc, 0xc00019ff80, 0x31, 0x7ca5a, ..
.)
        /restic/internal/archiver/scanner.go:97 +0x4a5
github.com/restic/restic/internal/archiver.(*Scanner).scan(0xc0005db320, 0x103ef00, 0xc0000b2200, 0x7ca5a, 0x1f851, 0x1b7f, 0xaaea63a34b, 0xc006e92180, 0x29, 0x76705, ..
.)
        /restic/internal/archiver/scanner.go:97 +0x4a5
github.com/restic/restic/internal/archiver.(*Scanner).scan(0xc0005db320, 0x103ef00, 0xc0000b2200, 0x76705, 0x1df7d, 0x1b70, 0xa4c05f98ce, 0xc00511e040, 0x1d, 0x75d9d, ..
.)
        /restic/internal/archiver/scanner.go:97 +0x4a5
github.com/restic/restic/internal/archiver.(*Scanner).scan(0xc0005db320, 0x103ef00, 0xc0000b2200, 0x75d9d, 0x1dca0, 0x1b70, 0xa4b3f18c7b, 0x7ffcc104adb0, 0x4, 0x75b45, .
..)
        /restic/internal/archiver/scanner.go:97 +0x4a5
github.com/restic/restic/internal/archiver.(*Scanner).Scan(0xc0005db320, 0x103ef00, 0xc0000b2200, 0xc0000b3780, 0x4, 0x4, 0x0, 0x0)
        /restic/internal/archiver/scanner.go:50 +0x16f
main.runBackup.func5(0x0, 0x0)
        /restic/cmd/restic/cmd_backup.go:576 +0x9a
gopkg.in/tomb%2ev2.(*Tomb).run(0xc0000cab40, 0xc00020a740)
        /home/build/go/pkg/mod/gopkg.in/tomb.v2@v2.0.0-20161208151619-d5d1b5820637/tomb.go:163 +0x38
created by gopkg.in/tomb%2ev2.(*Tomb).Go
        /home/build/go/pkg/mod/gopkg.in/tomb.v2@v2.0.0-20161208151619-d5d1b5820637/tomb.go:159 +0xbc

rax    0xca
rbx    0x158c440
rcx    0x46ce83
rdx    0x0
rdi    0x158c588
rsi    0x80
rbp    0x7ffcc1049230
rsp    0x7ffcc10491e8
r8     0x0
r9     0x0
r10    0x0
r11    0x286
r12    0x3
r13    0x158bb20
r14    0x80c000800000
r15    0x80c0009adfff
rip    0x46ce81
rflags 0x286
cs     0x33
fs     0x0
gs     0x0
```

**Comment by MichaelEischer** [comment]
Thanks for reporting this, the issue has already been fixed in #2963 . You can get a fixed restic version by either building the current master branch yourself or downloading a beta build from https://beta.restic.net/?sort=time&order=desc .

**Comment by andreasnuesslein** [comment]
Hello @MichaelEischer , I'm having the same problem but with `restic forget` - I tried the latest beta but it didn't change anything. Should the fix solve `forget` as well or is it possible that the same problem exists for `forget` and only `backup` has been fixed?

I can try and provide stacktraces and such if needed. 

Thank you!

**Comment by MichaelEischer** [comment]
@andreasnuesslein The fix from #2963 is only relevant for the `backup` command. When `forget` gets stuck somewhere, then that's a different bug. Please open a new issue for that and attach a stacktrace to it (send a SIGABRT or SIGQUIT signal to restic), thanks.

**Comment by andreasnuesslein** [comment]
@MichaelEischer I created a seperate ticket https://github.com/restic/restic/issues/3047

thank you :)
